### PR TITLE
Add framework for private messaging support

### PR DESCRIPTION
**Pull request by miterosan**
~~Solves #773~~
Please note that you can only view your private chats.
The required logic for sending private messages is commented out on the server side.

### CODE REVIEW

**Comment by smoogipoo** [comment]
How did you test this? The right-click dropdown on users in chat should have a "send message" item that opens up a new tab. There should be a testcase for this (along with normal chanels, but you removed those too?) that does not require api access.

Will continue review once there is a way to test this without having to... message myself?

File: osu.Game/Graphics/Containers/LinkFlowContainer.cs [hunk 1]
```diff
@@ -80,7 +80,9 @@ public void AddLink(string text, string url, LinkAction linkType = LinkAction.Ex
                                 game?.ShowBeatmapSet(setId);
                             break;
                         case LinkAction.OpenChannel:
-                            game?.OpenChannel(linkArgument);
```
Comment (peppy) [inline osu.Game/Graphics/Containers/LinkFlowContainer.cs]: why can't we continue using the game-level channel opening?

File: osu.Game/Online/API/Requests/GetUserMessagesRequest.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) 2007-2018 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using System.Collections.Generic;
+using osu.Framework.IO.Network;
+using osu.Game.Online.Chat;
+
+namespace osu.Game.Online.API.Requests
+{
+    public class GetUserMessagesRequest : APIRequest<List<Message>>
```
Comment (peppy) [inline osu.Game/Online/API/Requests/GetUserMessagesRequest.cs]: `GetMessagesPrivateRequest`

File: osu.Game/Online/API/Requests/GetChannelMessagesRequest.cs [hunk 3]
```diff
@@ -8,12 +8,12 @@
 
 namespace osu.Game.Online.API.Requests
 {
-    public class GetMessagesRequest : APIRequest<List<Message>>
```
Comment (peppy) [inline osu.Game/Online/API/Requests/GetChannelMessagesRequest.cs]: This doesn't need to change.

File: osu.Game/Online/API/Requests/GetUserMessagesRequest.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Online/API/Requests/GetUserMessagesRequest.cs]: I'm pretty sure most of this code is shared with `GetMessagesRequest`, so make a common abstract base class?

File: osu.Game/Online/Chat/ChatBase.cs [hunk 4]
```diff
@@ -1,44 +1,25 @@
...
 using System.Collections.Generic;
 using System.Linq;
-using Newtonsoft.Json;
 using osu.Framework.Configuration;
 using osu.Framework.Lists;
 
 namespace osu.Game.Online.Chat
 {
-    public class Channel
+    public abstract class ChatBase
```
Comment (peppy) [inline osu.Game/Online/Chat/ChatBase.cs]: this naming is wrong

File: osu.Game/Online/Chat/UserChat.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) 2007-2018 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using System;
+using osu.Game.Users;
+
+namespace osu.Game.Online.Chat
+{
+    public class UserChat : ChatBase
```
Comment (peppy) [inline osu.Game/Online/Chat/UserChat.cs]: Why can't this just derive from `Channel` (and be called `PrivateChannel`)?

File: osu.Game/OsuGame.cs [hunk 6]
```diff
@@ -142,12 +142,6 @@ private void load(FrameworkConfigManager frameworkConfig)
 
         private ScheduledDelegate scoreLoad;
 
-        /// <summary>
-        /// Open chat to a channel matching the provided name, if present.
-        /// </summary>
-        /// <param name="channelName">The name of the channel.</param>
-        public void OpenChannel(string channelName) => chat.OpenChannel(chat.AvailableChannels.Find(c => c.Name == channelName));
```
Comment (peppy) [inline osu.Game/OsuGame.cs]: why? at very least if this is moved it should find a new home in `ChatManager`.

File: osu.Game/Overlays/ChatOverlay.cs [hunk 7]
```diff
@@ -270,10 +351,9 @@ protected override void PopOut()
         }
 
         [BackgroundDependencyLoader]
-        private void load(APIAccess api, OsuConfigManager config, OsuColour colours)
+        private void load(APIAccess api, OsuConfigManager config, OsuColour colours, ChatManager chatManager)
         {
-            this.api = api;
-            api.Register(this);
+            api.Register(chatManager);
```
Comment (peppy) [inline osu.Game/Overlays/ChatOverlay.cs]: this is being registered in OsuGame and also here?

File: osu.Game/Online/Chat/ChatManager.cs [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+using osu.Framework.Threading;
+using osu.Game.Online.API;
+using osu.Game.Online.API.Requests;
+
+namespace osu.Game.Online.Chat
+{
+    /// <summary>
+    /// Manages everything chat related
+    /// </summary>
+    public sealed class ChatManager : IOnlineComponent
```
Comment (peppy) [inline osu.Game/Online/Chat/ChatManager.cs]: Should also be a `Component`, and added to the draw heirarchy (see `APIAccess`). You won't need to pass a `Scheduler` in any more.

File: osu.Game/Online/Chat/ChatManager.cs [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+        /// </summary>
+        private readonly string[] defaultChannels =
+        {
+            @"#lazer", @"#osu", @"#lobby"
+        };
+
+        /// <summary>
+        /// The currently opened chat
+        /// </summary>
+        public Bindable<ChatBase> CurrentChat { get; } = new Bindable<ChatBase>();
```
Comment (peppy) [inline osu.Game/Online/Chat/ChatManager.cs]: newline separation please

**Review by peppy — CHANGES_REQUESTED** [review]
A basic review at face value. Needs more tidying before I'm willing to dive in.

File: osu.Game/Online/Chat/ChatBase.cs (hunk 4 above)
Comment (miterosan) [reply osu.Game/Online/Chat/ChatBase.cs]: Just naming it chat would force me to use chat.chat everywhere

Comment (peppy) [reply osu.Game/Online/Chat/ChatBase.cs]: It's not `Chat`, it's a `Channel`

File: osu.Game/Online/Chat/UserChat.cs (hunk 5 above)
Comment (miterosan) [reply osu.Game/Online/Chat/UserChat.cs]: I didn't do that because then would userchannel be polluted with stuff it doesn't need. I named it userchat, because the player chats with another user or in a channel. (And user... because Chatbase.Targettype.User) 
Private channel sound more like a chat where more than 2 users are present. Like a private chatroom, or groupchat.

Comment (peppy) [reply osu.Game/Online/Chat/UserChat.cs]: Everything should be sharing a common `Channel` interface, because that is how they are server-side. None of the existing properties are polluting because they will be available in private chats too.

File: osu.Game/Online/Chat/Channel.cs [hunk 10]
```diff
@@ -22,27 +26,41 @@ public class Channel
...
-        public const int MAX_HISTORY = 300;
+        public long Id;
 
         [JsonConstructor]
         public Channel()
         {
         }
 
+        /// <summary>
+        /// Contructs a privatechannel
```
Comment (VPeruS) [inline osu.Game/Online/Chat/Channel.cs]: space

File: osu.Game/Online/Chat/Channel.cs [hunk 11]
```diff
@@ -22,27 +26,41 @@ public class Channel
...
+        /// TODO this class needs to be serialized from something like channels/private, instead of creating from a contructor
+        /// </summary>
+        /// <param name="user">The user </param>
+        public Channel(User user)
+        {
+            Target = TargetType.User;
+            Name = user.Username;
+            Id = user.Id;
```
Comment (VPeruS) [inline osu.Game/Online/Chat/Channel.cs]: This isn't supposed to be unique?

File: osu.Game.Tests/Visual/TestCaseChatTabControl.cs [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+namespace osu.Game.Tests.Visual
+{
+    public class TestCaseChatTabControl : OsuTestCase
+    {
+        public override IReadOnlyList<Type> RequiredTypes => new[]
+        {
+            typeof(ChatTabControl),
+            typeof(ChannelTabControl),
+            typeof(UserTabControl),
+
```
Comment (VPeruS) [inline osu.Game.Tests/Visual/TestCaseChatTabControl.cs]: newline and `,` at previous entry

File: osu.Game/Graphics/Containers/LinkFlowContainer.cs [hunk 13]
```diff
@@ -23,15 +23,15 @@ public LinkFlowContainer(Action<SpriteText> defaultCreationParameters = null)
         public override bool HandleMouseInput => true;
 
         private OsuGame game;
-
+        private ChannelManager chatManager;
```
Comment (VPeruS) [inline osu.Game/Graphics/Containers/LinkFlowContainer.cs]: Why not channelManager?

File: osu.Game/Online/Chat/ChannelManager.cs [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+        private void fetchNewMessages()
+        {
+            if (fetchChannelMsgReq == null)
+                fetchNewChannelMessages();
+
+            if (fetchUserMsgReq == null)
+                fetchNewUserMessages();
+        }
+
+        private void fetchNewUserMessages()
```
Comment (VPeruS) [inline osu.Game/Online/Chat/ChannelManager.cs]: looks very similar to fetchNewChannelMessages, you can try to create fetchMessages(APIRequest r)

Comment (miterosan) [reply osu.Game/Online/Chat/ChannelManager.cs]: ? Please explain

File: osu.Game/Overlays/ChatOverlay.cs [hunk 15]
```diff
@@ -285,253 +342,27 @@ private void load(APIAccess api, OsuConfigManager config, OsuColour colours)
...
-                IsAction = isAction,
-                Content = postText
-            };
-
-            var req = new PostMessageRequest(message);
-
-            target.AddLocalEcho(message);
-            req.Failure += e => target.ReplaceMessage(message, null);
-            req.Success += m => target.ReplaceMessage(message, m);
+            if (text[0] == '/')
```
[the remark this answers is not in the document]
Comment (miterosan) [reply osu.Game/Overlays/ChatOverlay.cs]: Unsure

File: osu.Game/Online/Chat/ChannelManager.cs (hunk 14 above)
Comment (VPeruS) [reply osu.Game/Online/Chat/ChannelManager.cs]:
```
fetchMessages(APIRequest requset, Action<IEnumerable<Message>> handler) { code from fetchNewUserMessages }

fetchMessages(new GetUserMessagesRequest(lastUserMsgId), handleUserMessages)
fetchMessages(new GetChannelMessagesRequest(JoinedChannels.Where(c => c.Target == TargetType.Channel), lastChannelMsgId)), handleChannelMessages)
```

**Comment by peppy** [comment]
This crashes on startup currently

**Comment by miterosan** [comment]
I can not replicate the startup crash.... https://drive.google.com/uc?id=165etj2zRhdJYN7Q3e9m-_Jk5QHtwTZfn

I fixed the exceptions in the testcase

File: osu.Game/Overlays/Chat/DrawableChat.cs [hunk 16]
```diff
@@ -15,15 +15,15 @@
 
 namespace osu.Game.Overlays.Chat
 {
-    public class DrawableChannel : Container
+    public class DrawableChat : Container
```
Comment (peppy) [inline osu.Game/Overlays/Chat/DrawableChat.cs]: why was this renamed?

File: osu.Game/Overlays/Chat/DrawableChat.cs [hunk 17]
```diff
@@ -71,14 +71,14 @@ protected override void Dispose(bool isDisposing)
...
-            Channel.PendingMessageResolved -= pendingMessageResolved;
+            Chat.NewMessagesArrived -= newMessagesArrived;
+            Chat.MessageRemoved -= messageRemoved;
+            Chat.PendingMessageResolved -= pendingMessageResolved;
         }
 
         private void newMessagesArrived(IEnumerable<Message> newMessages)
         {
-            // Add up to last Channel.MAX_HISTORY messages
+            // Add up to last ChatBase.MAX_HISTORY messages
```
Comment (peppy) [inline osu.Game/Overlays/Chat/DrawableChat.cs]: what is ChatBase?

File: osu.Game/Overlays/Chat/ChatTabItemCloseButton.cs [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+using osu.Framework.Graphics;
+using osu.Framework.Input;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Containers;
+using OpenTK;
+using OpenTK.Graphics;
+
+namespace osu.Game.Overlays.Chat
+{
+    public class ChatTabItemCloseButton : OsuClickableContainer
```
Comment (peppy) [inline osu.Game/Overlays/Chat/ChatTabItemCloseButton.cs]: Why is this in its own class when the consumers are private? I'd make everything public, and probably put them in a `Tabs` namespace.

File: osu.Game/Overlays/Chat/UserTabItem.cs [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+using osu.Game.Graphics.Sprites;
+using osu.Game.Online.Chat;
+using osu.Game.Screens.Menu;
+using osu.Game.Users;
+using OpenTK;
+using OpenTK.Graphics;
+
+namespace osu.Game.Overlays.Chat
+{
+    public class UserTabItem : TabItem<Channel>
```
Comment (peppy) [inline osu.Game/Overlays/Chat/UserTabItem.cs]: Why is this public (and in the top-level chat namespace) when `ChannelTabItem` isn't?

File: osu.Game/Overlays/Chat/ChatTabItemCloseButton.cs (hunk 18 above)
Comment (peppy) [reply osu.Game/Overlays/Chat/ChatTabItemCloseButton.cs]: Also why is there both a `ChatTabItemCloseButton` and a `CloseButton`?

File: osu.Game/Overlays/ChatOverlay.cs [hunk 20]
```diff
@@ -14,42 +13,36 @@
...
+    public class ChatOverlay : OsuFocusedOverlayContainer
     {
         private const float textbox_height = 60;
         private const float channel_selection_min_height = 0.3f;
 
-        private ScheduledDelegate messageRequest;
+        private ChannelManager channelManager;
 
-        private readonly Container<DrawableChannel> currentChannelContainer;
+        private readonly Container<DrawableChat> currentChatContainer;
```
Comment (peppy) [inline osu.Game/Overlays/ChatOverlay.cs]: Please rename all these "chat" changes back to "channel". It's just confusing.

File: osu.Game/Overlays/ChatOverlay.cs [hunk 21]
```diff
@@ -184,13 +177,90 @@ public ChatOverlay()
...
+                    channelManager.JoinedChannels.Add(channel);
+            };
+            channelSelection.OnRequestLeave = channel => channelManager.JoinedChannels.Remove(channel);
+        }
+
+        private void availableChannelsChanged(object sender, NotifyCollectionChangedEventArgs args)
+        {
+            channelSelection.Sections = new[]
+            {
+                new ChannelSection
```
Comment (peppy) [inline osu.Game/Overlays/ChatOverlay.cs]: This seems like the wrong place to do this. If anything, `AvailableChannels` should be passed in to `channelSelection` and the drawable construction should happen there.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
Note that this review is purely a code review. I have not run this PR for further testing.

File: osu.Game/Overlays/Chat/ChatLine.cs [hunk 22]
```diff
@@ -236,20 +237,24 @@ private class MessageSender : OsuClickableContainer, IHasContextMenu
...
+            private void load(UserProfileOverlay profile, ChannelManager chatManager)
             {
                 Action = () => profile?.ShowUser(sender);
+                startChatAction = () => chatManager?.OpenUserChannel(sender);
             }
 
             public MenuItem[] ContextMenuItems => new MenuItem[]
             {
                 new OsuMenuItem("View Profile", MenuItemType.Highlighted, Action),
+                new OsuMenuItem("Start Chat", MenuItemType.Highlighted, startChatAction),
```
Comment (peppy) [inline osu.Game/Overlays/Chat/ChatLine.cs]: Highlighted type should only be used for the default action.

**Comment by peppy** [comment]
- Tests are broken when switching between them:

```
Unhandled Exception: 
Unhandled Exception: osu.Framework.Allocation.RecursiveLoadException: Disposed Drawables may never get a parent and return to the scene graph.
Object name: 'ChannelManager'.
```

- User tab appear animation is broken (different animation on first click vs hide/unside)
- What is the weird symbol in the background on user tabs? It's supposed to be an "@".
- This PR breaks being able to send messages, both public and (potentially) private.

**Comment by miterosan** [comment]
^ It is not possible to send private message, because the required logic is commented out on the server side. See Line 120 https://github.com/ppy/osu-web/blob/master/app/Http/Controllers/API/ChatController.php

The problems with sending message to public channels is fixed now.

**Review by peppy — DISMISSED** [review]
part of review

File: osu.Game/Online/Chat/ChannelManager.cs [hunk 23]
```diff
@@ -0,0 +1,336 @@
...
+            CurrentChannel.Value = JoinedChannels.FirstOrDefault(c => c.Target == TargetType.User && c.Id == user.Id)
+                                   ?? new PrivateChannel { User = user };
+        }
+
+        public ChannelManager()
+        {
+            CurrentChannel.ValueChanged += currentChannelChanged;
+
+            channelMessagesHandler = new IncomingMessagesHandler();
+            channelMessagesHandler.CreateMessagesRequest = () => new GetMessagesRequest(JoinedChannels.Where(c => c.Target == TargetType.Channel), channelMessagesHandler.LastMessageId);
```
Comment (peppy) [inline osu.Game/Online/Chat/ChannelManager.cs]: pass `lastMessageId` as a parameter rather than self-referencing here.

**Comment by miterosan** [comment]
this is again ready for review

**Comment by ticotaco72** [comment]
@miterosan what's the state of this one

**Comment by peppy** [comment]
Please check the commit log for the state. It's an ongoing effort.

**Comment by peppy** [comment]
State of this: merged back into master but needs to new API changes to be reapplied. Should be done tomorrow.

File: osu.Game/Online/Chat/ChannelManager.cs [hunk 24]
```diff
@@ -0,0 +1,406 @@
...
+        /// The Channels the player has joined
+        /// </summary>
+        public ObservableCollection<Channel> JoinedChannels { get; } = new ObservableCollection<Channel>(); //todo: should be publicly readonly
+
+        /// <summary>
+        /// The channels available for the player to join
+        /// </summary>
+        public ObservableCollection<Channel> AvailableChannels { get; } = new ObservableCollection<Channel>(); //todo: should be publicly readonly
+
+        /*private readonly IncomingMessagesHandler privateMessagesHandler;*/
```
Comment (smoogipoo) [inline osu.Game/Online/Chat/ChannelManager.cs]:
```suggestion
```

File: osu.Game/Online/Chat/ChannelManager.cs [hunk 25]
```diff
@@ -0,0 +1,406 @@
...
+                // if this is a PM and the first message, we need to do a special request to create the PM channel
+                if (currentChannel.Type == ChannelType.PM && !currentChannel.Joined)
+                {
+                    var createNewPrivateMessageRequest = new CreateNewPrivateMessageRequest(currentChannel.Users.First(), message);
+                    createNewPrivateMessageRequest.Success += createRes =>
+                    {
+                        currentChannel.Id = createRes.ChannelID;
+                        currentChannel.ReplaceMessage(message, createRes.Message);
+                        dequeueAndRun();
+                    };
```
Comment (smoogipoo) [inline osu.Game/Online/Chat/ChannelManager.cs]:
```suggestion
                    };

```

File: osu.Game/Online/Chat/ChannelManager.cs [hunk 26]
```diff
@@ -0,0 +1,406 @@
...
+                    api.Queue(createNewPrivateMessageRequest);
+                    return;
+                }
+
+                var req = new PostMessageRequest(message);
+                req.Success += m =>
+                {
+                    currentChannel.ReplaceMessage(message, m);
+                    dequeueAndRun();
+                };
```
Comment (smoogipoo) [inline osu.Game/Online/Chat/ChannelManager.cs]:
```suggestion
                };

```

File: osu.Game/Online/Chat/ChannelManager.cs [hunk 27]
```diff
@@ -0,0 +1,406 @@
...
+                req.Failure += exception =>
+                {
+                    Logger.Error(exception, "Posting message failed.");
+                    currentChannel.ReplaceMessage(message, null);
+                    dequeueAndRun();
+                };
```
Comment (smoogipoo) [inline osu.Game/Online/Chat/ChannelManager.cs]:
```suggestion
                };

```

**Review by smoogipoo — DISMISSED** [review]
The textbox isn't focused when opening a new PM window.

**Review by smoogipoo — APPROVED** [review]
